lib/src/main/kotlin/StatefulSolver.kt

```diff
@@ -1,105 +1,82 @@
 
-        // Selection
-        val bestState = selectNode(root!!)
+    override var _root = createNode(null, null, mdp.initialState())
 
-        if (verbose) {
-            traceln("Selected:")
-            displayNode(bestState)
+    override fun selectNode(node: StateNode<TState, TAction>): StateNode<TState, TAction> {
+        // If the node is terminal, return it
+        if (mdp.isTerminal(node.state)) {
+            return node
         }
 
-        // Expansion
-        val expandedState = expandNode(bestState)
+        val exploredActions = node.exploredActions()
 
-        if (verbose) {
-            traceln("Expanding:")
-            displayNode(expandedState)
+        // This state has not been fully explored
+        if (node.validActions.size != exploredActions.size) {
+            return node
         }
 
-        // Simulation
-        val simulatedReward = simulateState(expandedState)
-
-        traceln("Simulated Reward: $simulatedReward")
+        // This state has been explored, select best action
+        var bestAction = exploredActions.first()
+        var bestActionScore : Double? = null
+
+        for (action in exploredActions) {
+            val childrenOfAction = node.getChildren(action)
+            val actionN = childrenOfAction.sumOf { c -> c.n }
+            val actionReward = childrenOfAction.sumOf { c -> c.reward }
+            val actionScore = calculateUCT(node.n, actionN, actionReward, explorationConstant)
+
+            if (bestActionScore == null || actionScore > bestActionScore) {
+                bestAction = action
+                bestActionScore = actionScore
+            }
+        }
 
-        // Update
-        updateNode(expandedState, simulatedReward)
-    }
+        val newState = mdp.transition(node.state, bestAction)
 
-    private fun updateNode(stateNode: StateNode<TAction, TState>, simulatedReward: Double) {
-        var currentStateNode = stateNode
-        var currentReward = simulatedReward
+        val actionState = node.getChildren(bestAction).firstOrNull { s -> s.state == newState }
+        // New state reached by an explored action
+                ?: return createNode(node, bestAction, newState)
 
-        while (true)
-        {
-            currentStateNode.maxReward = max(currentReward, currentStateNode.maxReward)
-            currentStateNode.reward += currentReward
-            currentStateNode.n++
 
-            val parentActionNode = currentStateNode.parentAction() ?: break
-            parentActionNode.reward += currentReward
-            parentActionNode.n++
+        // Existing state reached by an explored action
+        return selectNode(actionState)
+    }
 
-            currentStateNode = parentActionNode.parentState()
-            currentReward *= rewardDiscountFactor
+    override fun expandNode(node: StateNode<TState, TAction>): StateNode<TState, TAction> {
+        // If the node is terminal, return it
+        if (node.isTerminal) {
```

Comment:
Is `isTerminal` a property of the MDP State? Why is it now a property of StateNode?

Reply:
It is a property of the MDP state, but it's cached on the state node so this value isn't recalculated more than once.